File: src/public/js/domains/util/scriptdat/format/scripttxtparser.ts
import assert from 'assert';
import parse5 from 'parse5';
import { subWeaponNumbers } from '../../../model/randomizer/items';
import LMObject from '../data/LMObject';
import { LMChild, LMWorld } from '../data/Script';
import ShopItemsData from './ShopItemsData';

export function parseScriptTxt(txt: string) {
  const root: ReadonlyArray<any> = (
    (<any>parse5.parse(txt)).childNodes[0].childNodes[1].childNodes
  );
  const talks = (
    root
      .filter(x => x.tagName === 'talk')
      .map((x) => {
        assert.equal(x.childNodes.length, 1);
        assert.equal(x.childNodes[0].nodeName, '#text');
        assert.equal(x.childNodes[0].value[0], '\n');
        return (<string>x.childNodes[0].value).slice(1);
      })
  );
  assert.equal(ShopItemsData.parse(talks[252])[0].number, subWeaponNumbers.handScanner);
  assert.equal(ShopItemsData.parse(talks[252])[0].price, 20);
  assert.equal(ShopItemsData.parse(talks[252])[0].flag, 65279);
  assert.equal(ShopItemsData.parse(talks[252])[1].number, subWeaponNumbers.ammunition);
  assert.equal(ShopItemsData.parse(talks[252])[1].price, 500);
  assert.equal(ShopItemsData.parse(talks[252])[1].flag, 65279);
  assert.equal(ShopItemsData.parse(talks[252])[2].number, subWeaponNumbers.buckler);
  assert.equal(ShopItemsData.parse(talks[252])[2].price, 80);
  assert.equal(ShopItemsData.parse(talks[252])[2].flag, 697);
  const worlds = (
    root
      .filter(world => world.tagName === 'world')
      .map(world => ({
        value: parseAttrs(world.attrs)[0],
        fields: (
          (<ReadonlyArray<any>>world.childNodes)
            .filter(field => field.tagName === 'field')
            .map((field) => {
              const children = flatChildren(field);
              return {
                attrs: parseAttrs(field.attrs),
                children: (
                  children
                    .filter(child => (
                      child.tagName !== 'object'
                      && child.tagName !== 'map'
                      && child.nodeName !== '#text'
                    ))
                    .map(parseChild)
                ),
                objects: (
                  children
                    .filter(child => child.tagName === 'object')
                    .map(parseObject)
                ),
                maps: (
                  children
                    .filter(child => child.tagName === 'map')
                    .map((child) => {
                      const mapChildren = flatChildren(child);
                      return {
                        attrs: parseAttrs(child.attrs),
                        children: (
                          mapChildren
                            .filter(x => (
                              x.tagName !== 'object'
                              && x.nodeName !== '#text'
                            ))
                            .map(parseChild)
                        ),
                        objects: (
                          mapChildren
                            .filter(object => object.tagName === 'object')
                            .map(parseObject)
                        ),
                      };
                    })
                ),
              };
            })
        ),
      }))
  );

  assert.equal(talks.length, 905);
  assert.equal(worlds[0].fields[0].objects[0].starts[0].number, 99999);
  assert.equal(worlds[0].fields[0].maps[0].objects[5].starts[0].number, 58);
  return { talks, worlds };
}

function parseAttrs(attrs: ReadonlyArray<{ name: string; value: string }>) {
  assert.equal(attrs.length, 1);
  assert.equal(attrs[0].value, '');
  return attrs[0].name.split(',').map(Number);
}

function flatChildren(
  root: { childNodes: ReadonlyArray<{ tagName: string; childNodes?: ReadonlyArray<any> }> },
): ReadonlyArray<any> {
  return root.childNodes.map((x) => {
    if (x.childNodes == null) {
      return [x];
    }
    if (x.tagName === 'object' || x.tagName === 'map') {
      return [x];
    }
    return [x].concat(flatChildren(<{ childNodes: ReadonlyArray<any> }>x));
  }).reduce((p, c) => p.concat(c), []);
}

function parseChild(child: any): LMChild {
  return {
    name: child.tagName,
    attrs: parseAttrs(child.attrs),
  };
}

function parseObject(object: any) {
  const attrs = parseAttrs(object.attrs);
  return new LMObject(
    attrs[0],
    attrs[1],
    attrs[2],
    attrs[3],
    attrs[4],
    attrs[5],
    attrs[6],
    flatChildren(object)
      .filter(x => x.nodeName !== '#text')
      .map((x) => {
        const startAttrs = parseAttrs(x.attrs);
        return {
          number: startAttrs[0],
          value: Boolean(startAttrs[1]),
        };
      }),
  );
}

export function stringifyScriptTxt(
  talks: ReadonlyArray<string>,
  worlds: ReadonlyArray<LMWorld>,
) {
  return [
    talks.map(x => (
      `<TALK>\n${x}</TALK>\n`
    )).join(''),
    worlds.map(world => [
      `<WORLD ${world.value}>\n`,
      world.fields.map(field => [
        `<FIELD ${field.attrs.join(',')}>\n`,
        field.children.map(stringifyChild).join(''),
        field.objects.map(stringifyObject).join(''),
        field.maps.map(map => [
          `<MAP ${map.attrs.join(',')}>\n`,
          map.children.map(stringifyChild).join(''),
          map.objects.map(stringifyObject).join(''),
          `</MAP>\n`,
        ].join('')).join(''),
        `</FIELD>\n`,
      ].join('')).join(''),
      `</WORLD>\n`,
    ].join('')).join(''),
  ].join('');
}

function stringifyChild(child: LMChild) {
  return `<${child.name.toUpperCase()} ${child.attrs.join(',')}>\n`;
}

function stringifyObject(obj: LMObject) {
  return [
    `<OBJECT ${obj.number},${obj.x},${obj.y},${obj.op1},${obj.op2},${obj.op3},${obj.op4}>\n`,
    obj.starts.map(start => (
      `<START ${start.number},${Number(start.value)}>\n`
    )).join(''),
    `</OBJECT>\n`,
  ].join('');
}

File: src-tauri/src/util/scriptdat/format/scripttxtparser.rs
use std::fmt::Write;

use anyhow::{anyhow, Result};
use scraper::{node::Attributes, ElementRef, Html};

use crate::{
    randomizer::items::SubWeaponNumber,
    util::scriptdat::{
        data::{
            lm_object::{LMObject, LMStart},
            script::{LMChild, LMField, LMMap, LMWorld},
        },
        format::shop_items_data,
    },
};

pub fn parse_script_txt(text: &str) -> Result<(Vec<String>, Vec<LMWorld>)> {
    let parser = Html::parse_fragment(text);
    let root = parser.root_element().child_elements().collect::<Vec<_>>();
    // NOTE: scraper converts all tag names to lowercase
    let talks: Vec<_> = root
        .iter()
        .filter(|x| x.value().name() == "talk")
        .map(|x| {
            x.text()
                .collect::<String>()
                .trim_start_matches('\n')
                .to_owned()
        })
        .collect();
    if cfg!(debug_assertions) {
        let first_shop = shop_items_data::parse(&talks[252]);
        debug_assert_eq!(first_shop.0.number, SubWeaponNumber::HandScanner as i8);
        debug_assert_eq!(first_shop.0.price, 20);
        debug_assert_eq!(first_shop.0.flag, 65279);
        debug_assert_eq!(first_shop.1.number, SubWeaponNumber::Ammunition as i8);
        debug_assert_eq!(first_shop.1.price, 500);
        debug_assert_eq!(first_shop.1.flag, 65279);
        debug_assert_eq!(first_shop.2.number, SubWeaponNumber::Buckler as i8);
        debug_assert_eq!(first_shop.2.price, 80);
        debug_assert_eq!(first_shop.2.flag, 697);
    }
    let worlds: Vec<LMWorld> = root
        .iter()
        .filter(|world| world.value().name() == "world")
        .map(|world| {
            Ok(LMWorld {
                value: u8::try_from(parse_attrs(&world.value().attrs)?[0])?,
                fields: world
                    .child_elements()
                    .filter(|field| field.value().name() == "field")
                    .map(|field| {
                        let children = flat_children(field);
                        let mut attrs = parse_attrs(&field.value().attrs)?.into_iter();
                        Ok(LMField {
                            attrs: (
                                u8::try_from(attrs.next().ok_or(anyhow!("No attributes found"))?)?,
                                u8::try_from(attrs.next().ok_or(anyhow!("No attributes found"))?)?,
                                u8::try_from(attrs.next().ok_or(anyhow!("No attributes found"))?)?,
                                u8::try_from(attrs.next().ok_or(anyhow!("No attributes found"))?)?,
                                u8::try_from(attrs.next().ok_or(anyhow!("No attributes found"))?)?,
                            ),
                            children: children
                                .iter()
                                .filter(|child| {
                                    child.value().name() != "object"
                                        && child.value().name() != "map"
                                })
                                .map(|&child| parse_child(child))
                                .collect::<Result<_>>()?,
                            objects: children
                                .iter()
                                .filter(|child| child.value().name() == "object")
                                .map(|&child| parse_object(child))
                                .collect::<Result<_>>()?,
                            maps: children
                                .iter()
                                .filter(|child| child.value().name() == "map")
                                .map(|&child| {
                                    let attrs = parse_attrs(&child.value().attrs)?;
                                    let map_children = flat_children(child);
                                    Ok(LMMap {
                                        attrs: (
                                            u8::try_from(attrs[0])?,
                                            u8::try_from(attrs[1])?,
                                            u8::try_from(attrs[2])?,
                                        ),
                                        children: map_children
                                            .iter()
                                            .filter(|x| x.value().name() != "object")
                                            .map(|&child| parse_child(child))
                                            .collect::<Result<_>>()?,
                                        objects: map_children
                                            .iter()
                                            .filter(|object| object.value().name() == "object")
                                            .map(|&object| parse_object(object))
                                            .collect::<Result<_>>()?,
                                    })
                                })
                                .collect::<Result<_>>()?,
                        })
                    })
                    .collect::<Result<_>>()?,
            })
        })
        .collect::<Result<_>>()?;

    debug_assert_eq!(talks.len(), 905);
    debug_assert_eq!(worlds[0].fields[0].objects[0].starts[0].number, 99999);
    debug_assert_eq!(worlds[0].fields[0].maps[0].objects[5].starts[0].number, 58);
    Ok((talks, worlds))
}

fn parse_attrs(attrs: &Attributes) -> Result<Vec<i32>> {
    Ok(attrs
        .keys()
        .next()
        .ok_or(anyhow!("No attributes found"))?
        .local
        .split(',')
        .map(|x| x.parse::<i32>())
        .collect::<Result<_, _>>()?)
}

fn flat_children(root: ElementRef) -> Vec<ElementRef> {
    root.child_elements()
        .map(|x| {
            if x.child_elements().count() == 0 {
                return vec![x];
            }
            if x.value().name() == "object" || x.value().name() == "map" {
                return vec![x];
            }
            let mut vec = vec![x];
            vec.append(&mut flat_children(x));
            vec
        })
        .reduce(|mut p, mut c| {
            p.append(&mut c);
            p
        })
        .unwrap_or_default()
}

fn parse_child(child: ElementRef) -> Result<LMChild> {
    Ok(LMChild {
        name: child.value().name().to_owned(),
        attrs: parse_attrs(&child.value().attrs)?,
    })
}

fn parse_object(object: ElementRef) -> Result<LMObject> {
    let attrs = parse_attrs(&object.value().attrs)?;
    Ok(LMObject {
        number: u16::try_from(attrs[0])?,
        x: attrs[1],
        y: attrs[2],
        op1: attrs[3],
        op2: attrs[4],
        op3: attrs[5],
        op4: attrs[6],
        starts: flat_children(object)
            .iter()
            .map(|x| {
                let start_attrs = parse_attrs(&x.value().attrs)?;
                Ok(LMStart {
                    number: start_attrs[0],
                    value: start_attrs[1] != 0,
                })
            })
            .collect::<Result<_>>()?,
    })
}

pub fn stringify_script_txt(talks: &[String], worlds: &[LMWorld]) -> String {
    [
        talks.iter().fold(String::new(), |mut output, x| {
            write!(output, "<TALK>\n{x}</TALK>\n").unwrap();
            output
        }),
        worlds
            .iter()
            .map(|world| {
                [
                    format!("<WORLD {}>\n", world.value),
                    world
                        .fields
                        .iter()
                        .map(|field| {
                            [
                                format!(
                                    "<FIELD {},{},{},{},{}>\n",
                                    field.attrs.0,
                                    field.attrs.1,
                                    field.attrs.2,
                                    field.attrs.3,
                                    field.attrs.4
                                ),
                                field
                                    .children
                                    .iter()
                                    .map(stringify_child)
                                    .collect::<String>(),
                                field
                                    .objects
                                    .iter()
                                    .map(stringify_object)
                                    .collect::<String>(),
                                field
                                    .maps
                                    .iter()
                                    .map(|map| {
                                        [
                                            format!(
                                                "<MAP {},{},{}>\n",
                                                map.attrs.0, map.attrs.1, map.attrs.2,
                                            ),
                                            map.children
                                                .iter()
                                                .map(stringify_child)
                                                .collect::<String>(),
                                            map.objects
                                                .iter()
                                                .map(stringify_object)
                                                .collect::<String>(),
                                            "</MAP>\n".to_owned(),
                                        ]
                                        .join("")
                                    })
                                    .collect::<String>(),
                                "</FIELD>\n".to_owned(),
                            ]
                            .join("")
                        })
                        .collect::<String>(),
                    "</WORLD>\n".to_owned(),
                ]
                .join("")
            })
            .collect::<String>(),
    ]
    .join("")
}

fn stringify_child(child: &LMChild) -> String {
    format!(
        "<{} {}>\n",
        child.name.to_uppercase(),
        child
            .attrs
            .iter()
            .map(|x| x.to_string())
            .collect::<Vec<_>>()
            .join(",")
    )
}

fn stringify_object(object: &LMObject) -> String {
    format!(
        "<OBJECT {},{},{},{},{},{},{}>\n{}</OBJECT>\n",
        object.number,
        object.x,
        object.y,
        object.op1,
        object.op2,
        object.op3,
        object.op4,
        object
            .starts
            .iter()
            .fold(String::new(), |mut output, start| {
                writeln!(
                    output,
                    "<START {},{}>",
                    start.number,
                    if start.value { 1 } else { 0 }
                )
                .unwrap();
                output
            }),
    )
}
